File: crates/flowrt-codegen/tests/golden/sync_fusion_rust/expected/rust/src/selfdesc.rs
// FlowRT 管理产物。不要手工修改。

#[used]
#[unsafe(link_section = ".flowrt.selfdesc")]
static FLOWRT_SELF_DESCRIPTION: [u8; 11273] = *br#"{
  "self_description_version": "0.1",
  "ir_version": "0.1",
  "schema_version": "0.1",
  "source_hash": "806f2e9840ac7a04ae723f0b6c56755725c7f035d37693de4b6495c32ff2b82c",
  "artifact": {
    "mode": "strict",
    "temporary_island": false,
    "test_only": false,
    "clock": {
      "source": "realtime",
      "unit": "ms",
      "field": "tick_time_ms"
    }
  },
  "package": {
    "name": "sync_fusion_rust_demo",
    "version": null,
    "rsdl_version": "0.1"
  },
  "profiles": [
    {
      "name": "default",
      "mode": "strict",
      "backend": "inproc"
    }
  ],
  "targets": [
    {
      "name": "linux",
      "platform": null,
      "runtimes": [
        "rust"
      ],
      "backends": [
        "inproc"
      ]
    }
  ],
  "deployments": [
    {
      "graph": "default",
      "profile": "default",
      "target": "linux",
      "backend": "inproc",
      "satisfied": true
    }
  ],
  "graphs": [
    {
      "name": "default",
      "mode": "strict",
      "scheduler": {
        "worker_threads": 1,
        "lanes": [
          {
            "name": "fusion_serial",
            "kind": "serial",
            "instance": "fusion"
          },
          {
            "name": "imu_src_serial",
            "kind": "serial",
            "instance": "imu_src"
          },
          {
            "name": "odom_src_serial",
            "kind": "serial",
            "instance": "odom_src"
          },
          {
            "name": "sink_serial",
            "kind": "serial",
            "instance": "sink"
          }
        ],
        "tasks": [
          {
            "name": "main",
            "instance": "fusion",
            "lane": "fusion_serial",
            "trigger": "on_synchronized",
            "readiness": "any_ready",
            "concurrency": "exclusive",
            "period_ms": null,
            "deadline_ms": null,
            "priority": null
          },
          {
            "name": "main",
            "instance": "imu_src",
            "lane": "imu_src_serial",
            "trigger": "periodic",
            "readiness": "any_ready",
            "concurrency": "exclusive",
            "period_ms": 10,
            "deadline_ms": null,
            "priority": null
          },
          {
            "name": "main",
            "instance": "odom_src",
            "lane": "odom_src_serial",
            "trigger": "periodic",
            "readiness": "any_ready",
            "concurrency": "exclusive",
            "period_ms": 10,
            "deadline_ms": null,
            "priority": null
          },
          {
            "name": "main",
            "instance": "sink",
            "lane": "sink_serial",
            "trigger": "on_message",
            "readiness": "any_ready",
            "concurrency": "exclusive",
            "period_ms": null,
            "deadline_ms": null,
            "priority": null
          }
        ]
      },
      "resource_contract": {
        "resource_contract_version": "0.1",
        "requirements": [],
        "providers": [],
        "satisfactions": []
      },
      "external_processes": [],
      "instances": [
        {
          "name": "fusion",
          "component": "fusion",
          "process": "main",
          "target": null,
          "runtime": "rust",
          "params": []
        },
        {
          "name": "imu_src",
          "component": "imu_src",
          "process": "main",
          "target": null,
          "runtime": "rust",
          "params": []
        },
        {
          "name": "odom_src",
          "component": "odom_src",
          "process": "main",
          "target": null,
          "runtime": "rust",
          "params": []
        },
        {
          "name": "sink",
          "component": "sink",
          "process": "main",
          "target": null,
          "runtime": "rust",
          "params": []
        }
      ],
      "tasks": [
        {
          "name": "main",
          "instance": "fusion",
          "trigger": "on_synchronized",
          "readiness": "any_ready",
          "concurrency": "exclusive",
          "period_ms": null,
          "deadline_ms": null,
          "lane": "fusion_serial",
          "priority": null,
          "inputs": [],
          "outputs": [
            "estimate"
          ]
        },
        {
          "name": "main",
          "instance": "imu_src",
          "trigger": "periodic",
          "readiness": "any_ready",
          "concurrency": "exclusive",
          "period_ms": 10,
          "deadline_ms": null,
          "lane": "imu_src_serial",
          "priority": null,
          "inputs": [],
          "outputs": [
            "imu"
          ]
        },
        {
          "name": "main",
          "instance": "odom_src",
          "trigger": "periodic",
          "readiness": "any_ready",
          "concurrency": "exclusive",
          "period_ms": 10,
          "deadline_ms": null,
          "lane": "odom_src_serial",
          "priority": null,
          "inputs": [],
          "outputs": [
            "odom"
          ]
        },
        {
          "name": "main",
          "instance": "sink",
          "trigger": "on_message",
          "readiness": "any_ready",
          "concurrency": "exclusive",
          "period_ms": null,
          "deadline_ms": null,
          "lane": "sink_serial",
          "priority": null,
          "inputs": [
            "estimate"
          ],
          "outputs": []
        }
      ],
      "channels": [
        {
          "from": "fusion.estimate",
          "to": "sink.estimate",
          "message_type": "Estimate",
          "backend": "inproc",
          "thread_affinity": "send_safe",
          "service": null,
          "key_expr": null,
          "channel": "latest",
          "depth": 1,
          "overflow": "drop_oldest",
          "stale_policy": "warn",
          "max_age_ms": null
        },
        {
          "from": "imu_src.imu",
          "to": "fusion.imu",
          "message_type": "Imu",
          "backend": "inproc",
          "thread_affinity": "send_safe",
          "service": null,
          "key_expr": null,
          "channel": "latest",
          "depth": 1,
          "overflow": "drop_oldest",
          "stale_policy": "warn",
          "max_age_ms": null
        },
        {
          "from": "odom_src.odom",
          "to": "fusion.odom",
          "message_type": "Odom",
          "backend": "inproc",
          "thread_affinity": "send_safe",
          "service": null,
          "key_expr": null,
          "channel": "latest",
          "depth": 1,
          "overflow": "drop_oldest",
          "stale_policy": "warn",
          "max_age_ms": null
        }
      ],
      "boundary_endpoints": [],
      "services": [],
      "operations": []
    }
  ],
  "component_types": [
    {
      "name": "fusion",
      "language": "rust",
      "kind": "native",
      "resources": [],
      "io_boundary": null,
      "inputs": [
        {
          "name": "imu",
          "type": "Imu"
        },
        {
          "name": "odom",
          "type": "Odom"
        }
      ],
      "outputs": [
        {
          "name": "estimate",
          "type": "Estimate"
        }
      ],
      "service_clients": [],
      "service_servers": [],
      "operation_clients": [],
      "operation_servers": [],
      "params": []
    },
    {
      "name": "imu_src",
      "language": "rust",
      "kind": "native",
      "resources": [],
      "io_boundary": null,
      "inputs": [],
      "outputs": [
        {
          "name": "imu",
          "type": "Imu"
        }
      ],
      "service_clients": [],
      "service_servers": [],
      "operation_clients": [],
      "operation_servers": [],
      "params": []
    },
    {
      "name": "odom_src",
      "language": "rust",
      "kind": "native",
      "resources": [],
      "io_boundary": null,
      "inputs": [],
      "outputs": [
        {
          "name": "odom",
          "type": "Odom"
        }
      ],
      "service_clients": [],
      "service_servers": [],
      "operation_clients": [],
      "operation_servers": [],
      "params": []
    },
    {
      "name": "sink",
      "language": "rust",
      "kind": "native",
      "resources": [],
      "io_boundary": null,
      "inputs": [
        {
          "name": "estimate",
          "type": "Estimate"
        }
      ],
      "outputs": [],
      "service_clients": [],
      "service_servers": [],
      "operation_clients": [],
      "operation_servers": [],
      "params": []
    }
  ],
  "message_abi": [
    {
      "type_name": "Estimate",
      "size_bytes": 8,
      "align_bytes": 8,
      "empty": false,
      "fields": [
        {
          "name": "x",
          "type": "f64",
          "offset_bytes": 0,
          "size_bytes": 8,
          "align_bytes": 8
        }
      ]
    },
    {
      "type_name": "Imu",
      "size_bytes": 16,
      "align_bytes": 8,
      "empty": false,
      "fields": [
        {
          "name": "ax",
          "type": "f64",
          "offset_bytes": 0,
          "size_bytes": 8,
          "align_bytes": 8
        },
        {
          "name": "stamp_ns",
          "type": "u64",
          "offset_bytes": 8,
          "size_bytes": 8,
          "align_bytes": 8
        }
      ]
    },
    {
      "type_name": "Odom",
      "size_bytes": 16,
      "align_bytes": 8,
      "empty": false,
      "fields": [
        {
          "name": "vx",
          "type": "f64",
          "offset_bytes": 0,
          "size_bytes": 8,
          "align_bytes": 8
        },
        {
          "name": "stamp_ns",
          "type": "u64",
          "offset_bytes": 8,
          "size_bytes": 8,
          "align_bytes": 8
        }
      ]
    }
  ],
  "message_frames": [
    {
      "type_name": "Estimate",
      "encoding": "canonical_frame_v1",
      "header_size_bytes": 8,
      "max_size_bytes": 8,
      "variable": false,
      "fields": [
        {
          "name": "x",
          "type": "f64",
          "header_offset_bytes": 0,
          "header_size_bytes": 8,
          "tail_max_bytes": null
        }
      ]
    },
    {
      "type_name": "Imu",
      "encoding": "canonical_frame_v1",
      "header_size_bytes": 16,
      "max_size_bytes": 16,
      "variable": false,
      "fields": [
        {
          "name": "ax",
          "type": "f64",
          "header_offset_bytes": 0,
          "header_size_bytes": 8,
          "tail_max_bytes": null
        },
        {
          "name": "stamp_ns",
          "type": "u64",
          "header_offset_bytes": 8,
          "header_size_bytes": 8,
          "tail_max_bytes": null
        }
      ]
    },
    {
      "type_name": "Odom",
      "encoding": "canonical_frame_v1",
      "header_size_bytes": 16,
      "max_size_bytes": 16,
      "variable": false,
      "fields": [
        {
          "name": "vx",
          "type": "f64",
          "header_offset_bytes": 0,
          "header_size_bytes": 8,
          "tail_max_bytes": null
        },
        {
          "name": "stamp_ns",
          "type": "u64",
          "header_offset_bytes": 8,
          "header_size_bytes": 8,
          "tail_max_bytes": null
        }
      ]
    }
  ]
}
"#;

#[allow(dead_code)]
pub fn self_description_json() -> &'static str {
    std::str::from_utf8(&FLOWRT_SELF_DESCRIPTION).expect("generated FlowRT self-description is UTF-8")
}

#[allow(dead_code)]
pub fn self_description_hash() -> &'static str {
    "aeddc405c0dd52cad9e8fc60eae2615ad4540804ca16f327378c37c5c9e55a65"
}

File: crates/flowrt-codegen/tests/golden/bounded_operation_iox2_cpp/expected/cpp/src/selfdesc.cpp
// FlowRT 管理产物。不要手工修改。
#include "flowrt_app/selfdesc.hpp"

#include <string_view>

namespace flowrt_app {
namespace {

#if defined(__GNUC__) || defined(__clang__)
[[gnu::used, gnu::section(".flowrt.selfdesc")]]
#endif
const char kFlowrtSelfDescription[] = R"({
  "self_description_version": "0.1",
  "ir_version": "0.1",
  "schema_version": "0.1",
  "source_hash": "8e6bb88bbae216b02d4aae83e072fe7ffe1cc1a56d4b1a24ae579533cd38210b",
  "artifact": {
    "mode": "strict",
    "temporary_island": false,
    "test_only": false,
    "clock": {
      "source": "realtime",
      "unit": "ms",
      "field": "tick_time_ms"
    }
  },
  "package": {
    "name": "bounded_operation_iox2_cpp",
    "version": null,
    "rsdl_version": "0.1"
  },
  "profiles": [
    {
      "name": "default",
      "mode": "strict",
      "backend": "iox2"
    }
  ],
  "targets": [
    {
      "name": "linux",
      "platform": null,
      "runtimes": [
        "cpp"
      ],
      "backends": [
        "iox2"
      ]
    }
  ],
  "deployments": [
    {
      "graph": "default",
      "profile": "default",
      "target": "linux",
      "backend": "iox2",
      "satisfied": true
    }
  ],
  "graphs": [
    {
      "name": "default",
      "mode": "strict",
      "scheduler": {
        "worker_threads": 1,
        "lanes": [
          {
            "name": "controller_serial",
            "kind": "serial",
            "instance": "controller"
          },
          {
            "name": "navigator_serial",
            "kind": "serial",
            "instance": "navigator"
          }
        ],
        "tasks": [
          {
            "name": "main",
            "instance": "controller",
            "lane": "controller_serial",
            "trigger": "periodic",
            "readiness": "any_ready",
            "concurrency": "exclusive",
            "period_ms": 100,
            "deadline_ms": null,
            "priority": null
          },
          {
            "name": "main",
            "instance": "navigator",
            "lane": "navigator_serial",
            "trigger": "periodic",
            "readiness": "any_ready",
            "concurrency": "exclusive",
            "period_ms": 1000,
            "deadline_ms": null,
            "priority": null
          }
        ]
      },
      "resource_contract": {
        "resource_contract_version": "0.1",
        "requirements": [],
        "providers": [],
        "satisfactions": []
      },
      "external_processes": [],
      "instances": [
        {
          "name": "controller",
          "component": "controller",
          "process": "client_proc",
          "target": "linux",
          "runtime": "cpp",
          "params": []
        },
        {
          "name": "navigator",
          "component": "navigator",
          "process": "server_proc",
          "target": "linux",
          "runtime": "cpp",
          "params": []
        }
      ],
      "tasks": [
        {
          "name": "main",
          "instance": "controller",
          "trigger": "periodic",
          "readiness": "any_ready",
          "concurrency": "exclusive",
          "period_ms": 100,
          "deadline_ms": null,
          "lane": "controller_serial",
          "priority": null,
          "inputs": [],
          "outputs": []
        },
        {
          "name": "main",
          "instance": "navigator",
          "trigger": "periodic",
          "readiness": "any_ready",
          "concurrency": "exclusive",
          "period_ms": 1000,
          "deadline_ms": null,
          "lane": "navigator_serial",
          "priority": null,
          "inputs": [],
          "outputs": []
        }
      ],
      "channels": [],
      "boundary_endpoints": [],
      "services": [],
      "operations": [
        {
          "name": "controller.plan",
          "canonical_id": "operation_6a76898c2802c4e0",
          "client_instance": "controller",
          "client_port": "plan",
          "server_instance": "navigator",
          "server_port": "plan",
          "goal_type": "PlanGoal",
          "feedback_type": "PlanFeedback",
          "result_type": "PlanResult",
          "backend": "iox2",
          "backend_source": "profile_default",
          "goal_frame": {
            "message_type": "PlanGoal",
            "encoding": "canonical_frame_v1",
            "variable": true,
            "bounded": true,
            "max_size_bytes": 16,
            "iox2_slot_cap_bytes": null
          },
          "start_request_frame": {
            "message_type": "OperationStartRequest<PlanGoal>",
            "encoding": "canonical_frame_v1",
            "variable": true,
            "bounded": true,
            "max_size_bytes": 40,
            "iox2_slot_cap_bytes": 40
          },
          "timeout_ms": 5000,
          "concurrency": "reject",
          "preempt": "reject",
          "queue_depth": 4,
          "max_in_flight": 1,
          "feedback": "latest",
          "result_retention_ms": 60000,
          "lowering": {
            "start_service": "FlowRT/service/__flowrt_operation_controller_plan_start",
            "start_key_expr": "",
            "cancel_service": "FlowRT/service/__flowrt_operation_controller_plan_cancel",
            "cancel_key_expr": "",
            "status_service": "FlowRT/service/__flowrt_operation_controller_plan_status",
            "status_key_expr": "",
            "feedback_channel": "__flowrt_operation_controller_plan_feedback",
            "result_channel": "__flowrt_operation_controller_plan_result"
          }
        }
      ]
    }
  ],
  "component_types": [
    {
      "name": "controller",
      "language": "cpp",
      "kind": "native",
      "resources": [],
      "io_boundary": null,
      "inputs": [],
      "outputs": [],
      "service_clients": [],
      "service_servers": [],
      "operation_clients": [
        {
          "name": "plan",
          "goal_type": "PlanGoal",
          "feedback_type": "PlanFeedback",
          "result_type": "PlanResult"
        }
      ],
      "operation_servers": [],
      "params": []
    },
    {
      "name": "navigator",
      "language": "cpp",
      "kind": "native",
      "resources": [],
      "io_boundary": null,
      "inputs": [],
      "outputs": [],
      "service_clients": [],
      "service_servers": [],
      "operation_clients": [],
      "operation_servers": [
        {
          "name": "plan",
          "goal_type": "PlanGoal",
          "feedback_type": "PlanFeedback",
          "result_type": "PlanResult"
        }
      ],
      "params": []
    }
  ],
  "message_abi": [
    {
      "type_name": "PlanFeedback",
      "size_bytes": 4,
      "align_bytes": 4,
      "empty": false,
      "fields": [
        {
          "name": "progress",
          "type": "f32",
          "offset_bytes": 0,
          "size_bytes": 4,
          "align_bytes": 4
        }
      ]
    },
    {
      "type_name": "PlanResult",
      "size_bytes": 1,
      "align_bytes": 1,
      "empty": false,
      "fields": [
        {
          "name": "accepted",
          "type": "bool",
          "offset_bytes": 0,
          "size_bytes": 1,
          "align_bytes": 1
        }
      ]
    }
  ],
  "message_frames": [
    {
      "type_name": "PlanFeedback",
      "encoding": "canonical_frame_v1",
      "header_size_bytes": 4,
      "max_size_bytes": 4,
      "variable": false,
      "fields": [
        {
          "name": "progress",
          "type": "f32",
          "header_offset_bytes": 0,
          "header_size_bytes": 4,
          "tail_max_bytes": null
        }
      ]
    },
    {
      "type_name": "PlanGoal",
      "encoding": "canonical_frame_v1",
      "header_size_bytes": 8,
      "max_size_bytes": 16,
      "variable": true,
      "fields": [
        {
          "name": "target",
          "type": "string<max=8>",
          "header_offset_bytes": 0,
          "header_size_bytes": 8,
          "tail_max_bytes": 8
        }
      ]
    },
    {
      "type_name": "PlanResult",
      "encoding": "canonical_frame_v1",
      "header_size_bytes": 1,
      "max_size_bytes": 1,
      "variable": false,
      "fields": [
        {
          "name": "accepted",
          "type": "bool",
          "header_offset_bytes": 0,
          "header_size_bytes": 1,
          "tail_max_bytes": null
        }
      ]
    }
  ]
}
)";

const char kFlowrtSelfDescriptionHash[] = "0871d77ad09391c4a63c93f2f5ea95ef79e44f44f989feadac325fe1886cfec3";

}  // namespace

std::string_view self_description_json() noexcept {
    return std::string_view{kFlowrtSelfDescription, sizeof(kFlowrtSelfDescription) - 1};
}

std::string_view self_description_hash() noexcept {
    return std::string_view{kFlowrtSelfDescriptionHash, sizeof(kFlowrtSelfDescriptionHash) - 1};
}

}  // namespace flowrt_app
